### DISCUSSION TITLE
Data not immediately written to disk; possibility of data loss on crash or power outage?

### DISCUSSION THREAD

**Discussion by Qrox** [TriliumNext/Trilium#3715] [Q&A] [answered]
When I modify a note in Trilium and check the database files, usually it shows that the modification timestamp is unchanged. I don't know much about how a SQLite database works, so my question is: will trilium lose the data since the last write to the disk during a power outage or when it crashes? And is there a way to force it to write all changes to the disk? Thanks!

**Comment by XXXJetfireXXX** [comment]
Afaik, Trilium saves any changes you make to notes every second, so nope, there should not be such possibility. Probably the timestamp updates less frequently than the database itself.

**Reply by XXXJetfireXXX** [reply]
Or maybe it only updates when you stop editing and go to a different note.

**Reply by XXXJetfireXXX** [reply]
Though I just tested it and my "Modified" timestamp updated immediately after I made a change to a note.
![image](https://user-images.githubusercontent.com/43646178/225295727-6aaca76b-40ae-418d-994a-54094f2a14e1.png)

**Reply by Qrox** [reply]
I edited a note again and the "modified" timestamp of `document.db-wal` updated once, but after that it did not further update when I continued editing the note. The timestamp of `document.db` did not updated at all. Tried going to a different note and the timestamps did not update either. I'm on Windows 10 in case that's relevant.

> Probably the timestamp updates less frequently than the database itself.

That's a possiblity, [This Microsoft article](https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times?source=recommendations) mentions that "the only guarantee about a file time stamp is that the file time is correctly reflected when the handle that makes the change is closed." Although strangely when I viewed the properties of the Trilium log files, the modified timestamp was immediately updated with the latest value, but when I viewed the properties of the database files, the modified timestamp did not change, so I'm still not sure if the database files were actually updated.

**Comment by zadam** [comment, marked as answer]
As @XXXJetfireXXX changes are saved every second. The dateModified does not have to reflect real write progress. The database file is open for the whole process execution duration.

SQLite is quite resistant against power outages (depends also on underlying filesystem and hardware of course).

**Reply by Qrox** [reply]
You're right. I checked the checksum of the database files and `document.db-shm` and `document.db-wal` were indeed updated aftering editing notes.
